docs/architecture/adr-021-protobuf-query-encoding.md

```diff
@@ -0,0 +1,262 @@
+	return &balance, nil
+}
+```
+
+### Custom Query Registration and Routing
+
+Query server implementations as above would be registered with `AppModule`s using
+a new method `RegisterQueryServer(grpc.Server)` which could be implemented simply
+as below:
+
+```go
+// x/bank/module.go
+func (am AppModule) RegisterQueryServer(server grpc.Server) {
+	types.RegisterQueryServer(server, keeper.Querier{am.keeper})
+}
+```
+
+Underneath the hood, a new method `RegisterService(sd *grpc.ServiceDesc, handler interface{})`
+will be added to the existing `baseapp.QueryRouter` to add the queries to the custom
+query routing table (with the routing method being described below).
+The signature for this method matches the existing
+`RegisterServer` method on the GRPC `Server` type where `handler` is the custom
+query server implementation described above.
+
+
+GRPC-like requests are routed by the service name (ex. `cosmos_sdk.x.bank.v1.Query`)
+and method name (ex. `QueryBalance`) combined with `/`s to form a full
+method name (ex. `/cosmos_sdk.x.bank.v1.Query/QueryBalance`). This gets translated
+into an ABCI query as `custom/cosmos_sdk.x.bank.v1.Query/QueryBalance`. Service handlers
+registered with `QueryRouter.RegisterService` will be routed this way.
+
+Beyond the method name, GRPC requests carry a protobuf encoded payload, which maps naturally
+to `RequestQuery.Data`, and receive a protobuf encoded response or error. Thus
+there is a quite natural mapping of GRPC-like rpc methods to the existing
+`sdk.Query` and `QueryRouter` infrastructure.
+
+This basic specification allows us to reuse protocol buffer `service` definitions
+for ABCI custom queries substantially reducing the need for manual decoding and
+encoding in query methods. 
+
+### GRPC Protocol Support
+
+In addition to providing an ABCI query pathway, we can easily provide a GRPC
+proxy server that routes requests in the GRPC protocol to ABCI query requests
+under the hood. In this way, clients could use their host languages' existing
+GRPC implementations to make direct queries against Cosmos SDK app's using
+these `service` definitions. In order for this server to work, the `QueryRouter`
+on `BaseApp` will need to expose the service handlers registered with
+`QueryRouter.RegisterService` to the proxy server implementation. Nodes could
+launch the proxy server on a separate port in the same process as the ABCI app
+with a command-line flag.
+
+### REST Queries and Swagger Generation
+
+[grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway) is a project that
+translates REST calls into GRPC calls using special annotations on service
+methods. Apps that want to expose REST queries should add `google.api.http`
+annotations to their `rpc` methods as in this example below. By convention
+HTTP method paths registered this way should correspond to the protobuf
+package paths for their respective modules, i.e. `cosmos_sdk.x.bank.v1` translates
+to `/cosmos_sdk/x/bank/v1/`.
```

Comment:
While this may be the convention, I find this URI to not be extremely helpful. At the very least, we should consider omitting the `/comos_sdk` prefix.

Reply:
Well the argument for including it would be to set a precedent for other modules to namespace prefix all their paths. What do you think?

We could also take out the `/x`.

Reply:
What semantic "meaning" does the namespace have? e.g. `cosmos_sdk` doesn't really tell me much here.

Apps will most likely include modules from the SDK and maybe some of their own and some from other repos. Does that mean all the routes will have different prefixes?

Reply:
The namespace is just meant to be a prefix to prevent collision of modules from different repos. So I'm imagining every module would have both its short module name (ex. `bank`) and its repo/package name (ex. `cosmos_sdk`, `regen_modules`, etc.)

Reply:
I don't think that would provide a nice API. I agree we should prefix with the module name though.

Reply:
Can you propose an alternative?

One thing we should keep in mind is that swagger gets automatically generated so theoretically clients would never write out that URL, they would just use the swagger codegen where maybe that namespace would actually look sensible depending on the codegen (which I'll take a quick look at for js...)

Reply:
ohhh this is for swagger ONLY? As long as it's not part of the full URI, that's fine. Otherwise, I'd elect to omit it if possible.

Reply:
It's not for swagger only, it would be the full path. But if you want something shorter, can you suggest an alternative? Given that the way grpc-gateway works is by specifying the full REST path up front in `.proto` file (although that can be overidden by a separate yaml file I believe).

Reply:
Yes, by exactly omitting the first token. Instead of ``/cosmos_sdk/x/bank/v1/`, we have `/x/bank/v1/`.

Reply:
What would be the convention for other modules? Say Regen publishes the group module under the package name `regen_modules.group.v1`, we should probably do `/regen_modules/group/v1`?

Reply:
`x/group/v1`

Reply:
So no namespace prefix? What if there were conflicts between modules from different repos in the same app?

Reply:
If you think conflicts will not be an issue, I can just leave the example as `x/bank/v1` and not even mention conventions around REST paths.

Reply:
I do not think conflicts will be an issue. Why would you ever have two bank modules registered in your app? Or any two or more modules with the same name?

Reply:
I updated the ADR to reflect this. I'm fine going along with this, but my preference in general is for namespacing. I imagine if there were any conflicts down the road they could be addressed with some overrides via a yaml config.

Reply:
I would advise against namespacing that varies for a single cohesive RESTful API. Keep things simple, if it really becomes problem down the line, we can address it then.